server/src/internalClusterTest/java/org/elasticsearch/action/support/master/TransportMasterNodeActionIT.java

```diff
@@ -36,39 +39,49 @@ protected Collection<Class<? extends Plugin>> nodePlugins() {
     public void testRoutingLoopProtection() {
 
         final var newMaster = internalCluster().startMasterOnlyNode();
         final var enoughVotingMastersLatch = new CountDownLatch(1);
-        final var electionWonLatch = new CountDownLatch(1);
-        final var releaseMasterLatch = new CountDownLatch(1);
-
-        final ClusterStateApplier blockingApplier = event -> {
-            if (3 <= event.state().coordinationMetadata().getLastCommittedConfiguration().getNodeIds().size()) {
-                enoughVotingMastersLatch.countDown();
-            }
-
-            if (event.state().nodes().isLocalNodeElectedMaster()) {
-                logger.info("--> new master elected as planned");
-                electionWonLatch.countDown();
-                safeAwait(releaseMasterLatch);
-                logger.info("--> cluster state applications released on new master");
-            }
-        };
+        final var newMasterReceivedReroutedMessageLatch = new CountDownLatch(1);
 
         try {
-            internalCluster().getInstance(ClusterService.class, newMaster).addStateApplier(blockingApplier);
-
-            // need at least 3 voting master nodes for failover
+            /*
+             * Ensure that we've got 5 voting nodes in the cluster, this means even if the original
+             * master manages to accept its own failed state update before standing down, we can still
+             * establish a quorum without its (or our own) join.
+             */
+            internalCluster().getInstance(ClusterService.class, newMaster).addStateApplier(event -> {
+                if (5 <= event.state().coordinationMetadata().getLastCommittedConfiguration().getNodeIds().size()) {
+                    enoughVotingMastersLatch.countDown();
+                }
+            });
+            internalCluster().startMasterOnlyNode();
+            internalCluster().startMasterOnlyNode();
             internalCluster().startMasterOnlyNode();
             safeAwait(enoughVotingMastersLatch);
+            long originalTerm = internalCluster().masterClient().admin().cluster().prepareState().get().getState().term();
+
+            // Configure a latch that will be released when the existing master knows of the new master's election
+            final String originalMasterName = internalCluster().getMasterName();
+            logger.info("Original master was {}, new master will be {}", originalMasterName, newMaster);
+            final var previousMasterKnowsNewMasterIsElectedLatch = new CountDownLatch(1);
+            internalCluster().getInstance(ClusterService.class, originalMasterName).addStateApplier(event -> {
+                DiscoveryNode masterNode = event.state().nodes().getMasterNode();
+                if (masterNode != null && masterNode.getName().equals(newMaster)) {
+                    previousMasterKnowsNewMasterIsElectedLatch.countDown();
+                }
+            });
```

Comment:
because we're preventing cluster state updates being applied, we watch the previous master's cluster state to determine when we've won the election. We need to wait for the previous master to apply the term bump anyway before we send the redirected request so this serves multiple purposes.